crates/chia-datalayer/src/merkle.rs

```diff
@@ -0,0 +1,1902 @@
+    IndexIsNotAChild(TreeIndex),
+}
+
+// assumptions
+// - root is at index 0
+// - any case with no keys will have a zero length blob
+
+// define the serialized block format
+const METADATA_RANGE: Range<usize> = 0..METADATA_SIZE;
+const METADATA_SIZE: usize = 2;
+// TODO: figure out the real max better than trial and error?
+const DATA_SIZE: usize = 53;
+const BLOCK_SIZE: usize = METADATA_SIZE + DATA_SIZE;
+type BlockBytes = [u8; BLOCK_SIZE];
+type MetadataBytes = [u8; METADATA_SIZE];
+type DataBytes = [u8; DATA_SIZE];
+const DATA_RANGE: Range<usize> = METADATA_SIZE..METADATA_SIZE + DATA_SIZE;
+
+fn streamable_from_bytes_ignore_extra_bytes<T>(bytes: &[u8]) -> Result<T, Error>
+where
+    T: Streamable,
+{
+    let mut cursor = std::io::Cursor::new(bytes);
+    // TODO: consider trusted mode?
+    T::parse::<false>(&mut cursor).map_err(Error::Streaming)
+}
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Hash, Eq, PartialEq, Streamable)]
+pub enum NodeType {
+    Internal = 0,
+    Leaf = 1,
+}
+
+impl NodeType {
+    pub fn from_u8(value: u8) -> Result<Self, Error> {
+        streamable_from_bytes_ignore_extra_bytes(&[value])
+    }
+
+    #[allow(clippy::wrong_self_convention, clippy::trivially_copy_pass_by_ref)]
+    pub fn to_u8(&self) -> u8 {
+        Streamable::to_bytes(self).unwrap()[0]
+    }
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn sha256_num<T: ToBytes>(input: T) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(input.to_be_bytes());
+
+    Bytes32::new(hasher.finalize())
+}
+
+fn sha256_bytes(input: &[u8]) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(input);
+
+    Bytes32::new(hasher.finalize())
+}
+
+fn internal_hash(left_hash: &Hash, right_hash: &Hash) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(b"\x02");
+    hasher.update(left_hash);
+    hasher.update(right_hash);
+
+    Bytes32::new(hasher.finalize())
+}
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "Side", eq, eq_int))]
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Streamable)]
+pub enum Side {
+    Left = 0,
+    Right = 1,
+}
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "InsertLocation"))]
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+pub enum InsertLocation {
+    // error: Unit variant `Auto` is not yet supported in a complex enum
+    // = help: change to a struct variant with no fields: `Auto { }`
+    // = note: the enum is complex because of non-unit variant `Leaf`
+    Auto {},
+    AsRoot {},
+    Leaf { index: TreeIndex, side: Side },
+}
+
+#[derive(Copy, Clone, Hash, Debug, PartialEq, Eq, Streamable)]
+pub struct NodeMetadata {
+    // OPT: could save 1-2% of tree space by packing (and maybe don't do that)
+    pub node_type: NodeType,
+    pub dirty: bool,
+}
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "InternalNode", get_all))]
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, Streamable)]
+pub struct InternalNode {
+    parent: Parent,
```

Comment:
Maybe these fields should be public? Or are there invariants that need to be upheld - in that case, you could add getters on the Rust side instead.

I know this is for a Python binding but we should make sure the Rust API is solid too.

Reply:
made public in 0be9f5015b1854477cd3bd3633b09082efc89487